build.assets/tooling/cmd/resource-ref-generator/reference/reference.go

```diff
@@ -0,0 +1,154 @@
+type pageContent struct {
+	Resource resourceSection
+	Fields   map[resource.PackageInfo]resource.ReferenceEntry
+}
+
+// resourceSection represents a top-level section of the resource reference
+// dedicated to a dynamic resource.
+type resourceSection struct {
+	Version string
+	Kind    string
+	resource.ReferenceEntry
+}
+
+// TypeInfo represents the name and package name of an exported Go type. It
+// makes no guarantees about whether the type was actually declared within the
+// package.
+type TypeInfo struct {
+	// Go package path (not a file path)
+	Package string `yaml:"package"`
+	// Name of the type, e.g., Metadata
+	Name string `yaml:"name"`
+}
+
+// ResourceConfig describes a resource type to include in the reference.
+type ResourceConfig struct {
+	// The name of the struct type as declared in the Go source, e.g.,
+	// RoleV6.
+	TypeName string `yaml:"type"`
+	// The final path segment in the name of the Go package containing this
+	// type declaration, e.g., "api".
+	PackageName string `yaml:"package"`
+	// The value of the "kind" field within a YAML manifest for this
+	// resource, e.g., "role".
+	KindValue string `yaml:"yaml_kind"`
+	// The value of the "version" field within a YAML manifest for this
+	// resource, e.g., "v6".
+	VersionValue string `yaml:"yaml_version"`
+}
+
+// GeneratorConfig is the user-facing configuration for the resource reference
+// generator.
+type GeneratorConfig struct {
+	Resources []ResourceConfig `yaml:"resources"`
+	// Path to the root of the Go project directory.
+	SourcePath string `yaml:"source"`
+	// Directory where the generator writes reference pages.
+	DestinationDirectory string `yaml:"destination"`
+}
+
+type GenerationError struct {
+	messages []error
+}
+
+func (g GenerationError) Error() string {
+	// Begin with a newline to format the first list item below the outer
+	// error.
+	final := "\n"
+	for _, e := range g.messages {
+		final += fmt.Sprintf("- %v\n", e)
+	}
+	return final
+}
+
+// Generate uses the provided user-facing configuration to write the resource
+// reference to fs.
+func Generate(conf GeneratorConfig, tmpl *template.Template) error {
+	sourceData, err := resource.NewSourceData(conf.SourcePath)
+	if err != nil {
+		return fmt.Errorf("loading Go source files: %w", err)
+	}
+
+	var errs GenerationError
+	for _, r := range conf.Resources {
+		k := resource.PackageInfo{
+			DeclName:    r.TypeName,
+			PackageName: r.PackageName,
+		}
+
+		decl, ok := sourceData.TypeDecls[k]
+		if !ok {
+			errs.messages = append(errs.messages, fmt.Errorf("creating a reference entry for declaration %v.%v: cannot find a declaration of this resource type", k.PackageName, k.DeclName))
+			continue
+		}
+
+		pc := pageContent{}
+
+		// decl is a dynamic resource type, so get data for the type and
+		// its dependencies.
+		entries, err := resource.ReferenceDataFromDeclaration(decl, sourceData.TypeDecls)
+		if errors.Is(err, resource.NotAGenDeclError{}) {
```

Comment:
I would test this and make sure it does what you think. You might need to use `errors.As` here instead.